c/rc_lock.c

```diff
@@ -0,0 +1,464 @@
+
+  uint8_t proof_mask = 0x3;  // both
+  for (uint32_t i = 0; i < proof_len; i++) {
+    bool existing = false;
+    SmtProofEntryType proof_entry = proofs->t->get(proofs, i, &existing);
+    CHECK2(existing, ERROR_INVALID_MOL_FORMAT);
+    mol2_cursor_t proof = proof_entry.t->proof(&proof_entry);
+
+    const RCRule *current_rule = &rce_state->rcrules[i];
+    err = rce_verify_one_rule(rce_state, &states, NULL, NULL, proof_mask, proof,
+                              current_rule);
+    CHECK(err);
+  }
+  if (rce_state->has_wl) {
+    if (rce_state->both_on_wl) {
+      err = 0;
+    } else {
+      err = ERROR_NOT_ON_WHITE_LIST;
+    }
+  } else {
+    // all black list, it's not allowed
+    err = ERROR_NO_WHITE_LIST;
+  }
+exit:
+  return err;
+}
+
+#ifdef CKB_USE_SIM
+int simulator_main() {
+#else
+int main() {
+#endif
+  // parse lock script's args
+  int err = 0;
+  ArgsType args = {0};
+
+  err = parse_args(&args);
+  CHECK(err);
+
+  // extra signature from witness
+  WitnessArgsType witness;
+  err = make_witness(&witness);
+  CHECK(err);
+
+  BytesOptType lock = witness.t->lock(&witness);
+  CHECK2(!lock.t->is_none(&lock), ERROR_INVALID_MOL_FORMAT);
+  mol2_cursor_t lock_bytes = lock.t->unwrap(&lock);
+  // convert Bytes to RcLockWitnessLock
+  RcLockWitnessLockType witness_lock = make_RcLockWitnessLock(&lock_bytes);
+
+  // pubkey hash
+  if (args.id.flags == IdentityFlagsPubkeyHash ||
+      args.id.flags == IdentityFlagsPubkeyHashRc) {
+    uint8_t signature_bytes[SIGNATURE_SIZE];
+    BytesOptType signature_opt = witness_lock.t->signature(&witness_lock);
+    mol2_cursor_t signature_cursor = signature_opt.t->unwrap(&signature_opt);
+
+    uint32_t read_len =
+        mol2_read_at(&signature_cursor, signature_bytes, SIGNATURE_SIZE);
+    CHECK2(read_len == SIGNATURE_SIZE, ERROR_INVALID_MOL_FORMAT);
+
+    err = verify_secp256k1_blake160_sighash_all(args.id.blake160,
+                                                signature_bytes);
+    CHECK(err);
+  }
+
+  // owner lock
+  if (args.id.flags == IdentityFlagsOwnerLock ||
+      args.id.flags == IdentityFlagsOwnerLockRc) {
+    bool existing = is_lock_script_hash_existing(args.id.blake160);
+    CHECK2(existing, ERROR_LOCK_SCRIPT_HASH_NOT_FOUND);
+  }
+
+  // regulation compliance
+  if (args.id.flags == IdentityFlagsPubkeyHashRc ||
+      args.id.flags == IdentityFlagsOwnerLockRc) {
+    // collect rc rules
+    RceState rce_state;
+    rce_init_state(&rce_state);
+    err = rce_gather_rcrules_recursively(&rce_state, args.rc_root, 0);
+    CHECK(err);
+    CHECK2(rce_state.rcrules_count > 0, ERROR_NO_RCRULE);
+    CHECK2(rce_state.has_wl, ERROR_NO_WHITE_LIST);
+
+    // collect proof
+    SmtProofEntryVecOptType proofs_opt = witness_lock.t->proofs(&witness_lock);
+    SmtProofEntryVecType proofs = proofs_opt.t->unwrap(&proofs_opt);
+
+    // verify blake160 against proof, using rc rules
+    err = verify_identity(&args.id, &proofs, &rce_state);
```

Comment:
In case of RC unlocking, the identity should be provided in witness, we should not read this from `args.id`